src/main/java/org/dependencytrack/tasks/OsvDownloadTask.java

```diff
@@ -109,92 +213,462 @@ public OsvDownloadTask() {
+                + "-modified.csv.ts").getAbsoluteFile();
+        File fullTimestampFile = new File(outputDir, OSV_FILENAME_PREFIX + ecosystem
+                + ".zip.ts").getAbsoluteFile();
+        if (incremental) {
+            Path modifiedCsv = DOWNLOAD_RETRY.executeCheckedSupplier(() -> downloadModifiedCsvFile(url, ecosystem));
+            LOGGER.debug("Downloaded list of new or modified OSV advisories for " + ecosystem + " into " + modifiedCsv);
+            final boolean success = processModifiedCsvFile(modifiedCsv, ecosystem);
+            if (success) {
+                LOGGER.info("Incremental update completed for " + ecosystem);
+                writeTimestampFile(incrementalTimestampFile, startTime);
+            }
+        } else {
+            Path osvZipFile = DOWNLOAD_RETRY.executeCheckedSupplier(() -> downloadOsvZipFile(url, ecosystem));
+            LOGGER.debug("Downloaded OSV advisories for " + ecosystem + " into " + osvZipFile);
+            final boolean success = processOsvZipFile(osvZipFile);
+            if (success) {
+                LOGGER.info("Full mirror completed for " + ecosystem);
+                writeTimestampFile(fullTimestampFile, startTime);
+                writeTimestampFile(incrementalTimestampFile, startTime);
+            }
         }
+    }
 
-        for (final var ecosystem : this.ecosystems) {
-            LOGGER.info("Updating datasource with Google OSV advisories for ecosystem " + ecosystem);
-            final var url = this.osvBaseUrl + URLEncoder.encode(ecosystem, StandardCharsets.UTF_8).replace("+", "%20") + "/all.zip";
-            final var request = new HttpGet(url);
-            try (final var ignoredMdcOsvEcosystem = MDC.putCloseable("osvEcosystem", ecosystem); final var response = HttpClientPool.getClient().execute(request)) {
-                final var status = response.getStatusLine();
+    private Path downloadModifiedCsvFile(String url, String ecosystem) throws Throwable {
+        final long downloadStart = System.currentTimeMillis();
+        final HttpUriRequest request = new HttpGet(url);
+        try (CloseableHttpResponse response = REQUEST_RETRY.executeCheckedSupplier(() ->
+                HttpClientPool.getClient().execute(request))) {
+            final StatusLine status = response.getStatusLine();
+            final HttpEntity entity = response.getEntity();
+            try {
+                LOGGER.info("Downloading...");
                 if (status.getStatusCode() != HttpStatus.SC_OK) {
-                    LOGGER.error("Download failed : " + status.getStatusCode() + ": " + status.getReasonPhrase());
-                    continue;
+                    LOGGER.error("Download of modified_id.csv failed for: " + ecosystem + " - " +
+                            status.getStatusCode() + " " + status.getReasonPhrase());
+                    mirroredWithoutErrors = false;
+                    throw new Exception("Download failed: " + status);
```

Comment:
Don't use generic `Exception` or `RuntimeException`:

```suggestion
                    throw new IllegalStateException("Download failed: " + status);
```